clash-lib/src/Clash/Backend/VHDL.hs

```diff
@@ -2081,3 +1830,335 @@ tupledSemi :: Applicative f => f [Doc] -> f Doc
 tupledSemi = align . encloseSep (flatAlt (lparen <+> emptyDoc) lparen)
                                 (flatAlt (emptyDoc <+> rparen) rparen)
                                 (semi <+> emptyDoc)
+
+-- | VHDL name modifiers
+data VHDLModifier
+  -- | SLV slice (descending index)
+  = Range    Range
+  -- | Element selection
+  | Idx      Int
+  -- | Array slice (ascending index)
+  | Slice    Int Int
+  -- | Selected names
+  | Select   (VHDLM Doc)
+  -- | Projecting a 'Word#' out of a 'Word8', or 'Int#' ouf of an 'Int8', see
+  -- [Note] integer projection
+  | Resize
+  -- | Projecting a 'Natural' out of a 'BitVector', see [Note] bitvector projection
+  | ResizeAndConvert
+  -- | Projecting the mask out of a 'BitVector', see [Note] mask projection
+  | DontCare
+
+-- | Create a sequence of VHDL name modifiers from our internal 'Modifier'
+-- data type. Note that the modifiers are in "reverse" order, so build a
+-- complete modified name using 'foldr' over the list by this function.
+--
+-- [Note] Continuing from an SLV slice
+-- SOP and custom products are represented as std_logic_vector, this means that
+-- their elements are also std_logic_vector. So when we project an element out
+-- of an SOP or custom project, and want to do a further projection on that,
+-- we have to do further SLV slicing; instead of e.g. creating a 'selected'
+-- modifier. Finally, when we render the modified name, we have to check
+-- whether the ultimately projected type needs to be converted from this SLV
+-- slice, to the proper type.
+buildModifier
+  :: [(VHDLModifier,HWType)]
+  -- ^ The list of modifiers so far, note that this list is in reverse order
+  -- in which they should eventually be applied to the name we want to modify
+  -> Modifier
+  -> Maybe [(VHDLModifier,HWType)]
+  -- ^ 'Nothing' indicates that the 'Modifier' does not result into a VHDL name
+  -- modifier.
+buildModifier prevM (Sliced (_,start,end)) = case prevM of
+  (prev:rest)
+    | (Range r,_) <- prev -> -- See [Note] Continuing from an SLV slice
+      Just (first Range (continueWithRange [(start,end)] hty r) : rest)
+  _ ->
+      Just ((Range (Contiguous start end),hty) : prevM)
+ where
+  hty = BitVector (start-end+1)
+
+buildModifier prevM (Indexed (ty@(SP _ args),dcI,fI)) = case prevM of
+  (prev:rest)
+    | (Range r,_) <- prev -> -- See [Note] Continuing from an SLV slice
+      Just (first Range (continueWithRange [(start,end)] argTy r) : rest)
+  _ ->
+      Just ((Range (Contiguous start end),argTy) : prevM)
+ where
+  argTys   = snd $ args !! dcI
+  argTy    = argTys !! fI
```

Comment:
Might be best to use `indexNote` and annotate the function with `HasCallStack`.

Reply:
Same for the other uses of `!!`. Create an `indexField` / `indexCon` to not pollute the code?